Introduce `spec.free` to catalog restrictions for Plans

jeremyrickard opened the pull request:
This PR extends the existing catalog restrictions to allow the user to specify `spec.free` when providing [Cluster]ServicePlan catalog restrictions. This also adds relevant documentation for catalog restrictions. 

Closes: #2209, #2156

jeremyrickard commented:
Here is the docs preview: https://deploy-preview-2211--svc-cat.netlify.com/

jeremyrickard commented on docs/catalog-restrictions.md:
```diff
@@ -0,0 +1,194 @@
+
+This example creates a Service Class restriction on spec.externalName using the
+ `in` operator. In this case, only services that have the externalName 
+ `FooService` or `BarService` will have Service Catalog resources created. 
```
that's a good question! let me read the code and figure that out. Will add relevant text to this to make sure!

jboyd01 commented on docs/catalog-restrictions.md:
```diff
@@ -0,0 +1,194 @@
+In this example, a catalog restriction has been defined that specifies that 
+only service plans that have an external name of basic should be selected. 
+Catalog restrictions are defined as a set of one or more rules that target 
+either service classes and service plans. These rules have a special format
```
> Catalog restrictions are defined as a set of one or more rules that target either service classes and service plans

Because of the word *either*, I think *and* should be an *or*.  Perhaps we should say "that target service classes and/or service plans"

jeremyrickard replied:
Ahh yeah thanks!  Forgot to remove that word. You can supply both. so and/or

jboyd01 commented on docs/catalog-restrictions.md:
```diff
@@ -0,0 +1,194 @@
+| spec.externalID | This key will match the ServicePlan.Spec.ExternalID property |
+| spec.free | This key will match the ServicePlan.Spec.Free property |
+| spec.serviceClass.name | This key will match the ServicePlan.Spec.ServiceClassRef.Name property |
+
```
Since the property key and descriptions are the same for the NS and Cluster scoped Plan/Class, I'd suggest combining them.  IE  "`ClusterServicePlan` and `ServicePlan` allowed property names"
Same with the Classes

jeremyrickard replied:
They are _almost_ the same. There are some shared keys, but we also allow to supplies predicates that reference the non-common parts, i.e. ServicePlans can be filtered with ServiceClassRef.Name and ClusterServicePlans can be filtered with ClusterServiceClassRef.Name.

Would you favor doing a shared block and then call out the two differences?

jboyd01 replied:
@jeremyrickard I missed that.  I totally agree, leave it as is, thank you.

jeremyrickard commented:
@carolynvs @jboyd01 I've updated this PR to add this doc to the README.md under topics for operators and I modified the namespaced doc to link here, as was suggested in that PR. Take a 👀

jboyd01 commented:
/lgtm
